hedy.py

```diff
@@ -205,8 +205,12 @@ def __init__(self, **arguments):
         super().__init__('Too Big', **arguments)
 
 class InvalidCommandException(HedyException):
-    def __init__(self, **arguments):
-        super().__init__('Invalid', **arguments)
+    def __init__(self, level, location, invalid_command, guessed_command):
+        super().__init__('Invalid')
+        self.level = level
+        self.location = location
+        self.invalid_command = invalid_command
+        self.guessed_command = guessed_command
 
```

Comment:
If you place the arguments in a dictionary you do not have to set individual fields because the ** will take care of that for you.

Reply:
I'm not opposed to making the expected arguments explicit in the constructor signature. It will make it harder to misuse the class later on. For example, if you did the following:

```python
InvalidCommandException(invalid_commadn='xyz')
```

Then:

* The linter would see that you misspelled 'command'
* The linter would see that you are missing two arguments.

However, what happens to the arguments here is *different* than what used to happen:

## Old

```python
    def __init__(self, level, **arguments):
        super().__init__('Invalid', **arguments)
```

This passes all keyword arguments in `arguments` up to the **superclasss**.

##  New

```python
    def __init__(self, level, location, invalid_command, guessed_command):
        super().__init__('Invalid')
        self.level = level
        self.location = location
        self.invalid_command = invalid_command
        self.guessed_command = guessed_command
```

This doesn't pass any arguments up to the superclass anymore. Instead, it stores all arguments onto the current class.

## What's the difference

The mechanism where all exception parameters are passed up into the superclass (where they get stored into the `ex.arguments` dict) is used to inject these strings into formatted error messages.

So everything you don't pass to the superclass cannot be used in error message format strings! 

At the same time, `hedy_error_to_response()` will look at attributes that exist on the exception object itself (specifically `location`), to format them into the response object.

So we need to pass certain attributes up to the superclass (to be packed into `ex.arguments`), as well as store some onto the exception object directly.

## Recommendation

I would recommend writing it like this:

```python
    def __init__(self, *, level, location, invalid_command, guessed_command):
        super().__init__('Invalid', level=level, invalid_command=invalid_command, guessed_command=guessed_command)
        # Mark the error location for rendering in the frontend
        self.location = location
        self.location_length = invalid_command.length
```

This will have the following benefits:

* It makes the required arguments explicit, so the class cannot be accidentally misused
* The lone `*` in the argument list makes it so that all arguments MUST be passed as keyword arguments, not positionally. This makes it harder to accidentally mess up the order of the arguments.
* It still passes all the arguments that are likely to be used in the error format string up to the superclass
* It sets `location` and `location_length` to the values that are expected by `hedy_error_to_response`.